### PR TITLE
fix(api-gateway): make sure DAP works with sql pushdown

### PR DESCRIPTION
**Pull request by bsod90**
fixes an issue when processing queries with member expressions

**Check List**
- [ ] Tests has been run in packages where changes made if available
- [ ] Linter has been run for changed code
- [ ] Tests for the changes have been added if not covered yet
- [ ] Docs have been added / updated if required

**Issue Reference this PR resolves**

[For example #12]

**Description of Changes Made (if issue reference is not provided)**

[Description goes here]

### CODE REVIEW

File: packages/cubejs-api-gateway/src/gateway.ts [hunk 1]
```diff
@@ -1207,8 +1212,8 @@ class ApiGateway {
             context,
           );
 
-          if (hasExpressionsInQuery) {
-            rewrittenQuery = this.evalMemberExpressionsInQuery(rewrittenQuery);
+          if (this.hasExpressionsInQuery(rewrittenQuery)) {
+            rewrittenQuery = this.parseMemberExpressionsInQuery(rewrittenQuery);
```
Comment (mcheshkov) [inline packages/cubejs-api-gateway/src/gateway.ts:1216]: I've not yet checked yet what can `applyRowLevelSecurity` introduce

But `parseMemberExpressionsInQuery` is not enough anyway.
There's three states of member expressions:

1. `MemberExpression`
Object with `expression` method, that takes cubes as symbols and returns SQL

2. `ParsedMemberExpression`
Object with `expression` as array of strings. Essentially JSON-compatible way of storing `MemberExpression`

3. Just a string with JSON-serialzed `ParsedMemberExpression`.

Schema compiler expects only `MemberExpression`.
`parseMemberExpressionsInQuery` turns strings in proper places in query to `ParsedMemberExpression`.
`evalMemberExpressionsInQuery` turns `ParsedMemberExpression` to `MemberExpression` in those places.

So, depending on what `applyRowLevelSecurity` can add either only `evalMemberExpressionsInQuery`, or both of `parseMemberExpressionsInQuery` + `evalMemberExpressionsInQuery` are necessary.

File: packages/cubejs-api-gateway/src/gateway.ts [hunk 2]
```diff
@@ -1195,7 +1195,12 @@ class ApiGateway {
             currentQuery = this.parseMemberExpressionsInQuery(currentQuery);
           }
 
-          const normalizedQuery = normalizeQuery(currentQuery, persistent);
+          let normalizedQuery = normalizeQuery(currentQuery, persistent);
+
+          if (hasExpressionsInQuery) {
```
Comment (mcheshkov) [inline packages/cubejs-api-gateway/src/gateway.ts:1200]: Please, add comment about why early/second call is necessary. I'll probably forget about this whole issue next week :sweat_smile:

**Review by mcheshkov — CHANGES_REQUESTED** [review]

File: packages/cubejs-api-gateway/src/gateway.ts (hunk 1 above)
Comment (bsod90) [reply packages/cubejs-api-gateway/src/gateway.ts:1216]: `applyRowLevelSecurity` only adds `MemberExpression` objects and only in one special case: when the entire query should be denied, we add `WHERE 1 == 0`

Comment (bsod90) [reply packages/cubejs-api-gateway/src/gateway.ts:1216]: Correction: we actually add string member expressions: [
](https://github.com/cube-js/cube/blob/master/packages/cubejs-server-core/src/core/CompilerApi.js#L320)

Comment (bsod90) [reply packages/cubejs-api-gateway/src/gateway.ts:1216]: https://github.com/cube-js/cube/blob/master/packages/cubejs-server-core/src/core/CompilerApi.js#L320

Comment (mcheshkov) [reply packages/cubejs-api-gateway/src/gateway.ts:1216]: `expression: () => '1 = 0',`
That's `MemberExpression`, with `expression` as method, returnung SQL

String member expression would look like this: `query.segments.push('{"cubeName":"foo","expr": "1=0"}');`
So, for this case no need to add any additional calls after `applyRowLevelSecurity`

Comment (bsod90) [reply packages/cubejs-api-gateway/src/gateway.ts:1216]: Let me see if I can just add it as a parsed version and avoid this messy logic altogether

Comment (bsod90) [reply packages/cubejs-api-gateway/src/gateway.ts:1221]: Sorry, I've removed all my previous comment and just added the `eval`. Looks like it should be fine now

**Review by mcheshkov — APPROVED** [review]

**Pull request merged**